gbus/tx/mysql/timeout.go

```diff
@@ -0,0 +1,223 @@
+package mysql
+
+import (
+	"database/sql"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/wework/grabbit/gbus"
+)
+
+var _ gbus.TimeoutManager = &TimeoutManager{}
+
+//TimeoutManager is a mysql implementation of a persistent timeoutmanager
+type TimeoutManager struct {
+	Bus               gbus.Bus
+	Log               func() logrus.FieldLogger
+	TimeoutSaga       func(*sql.Tx, string) error
+	Txp               gbus.TxProvider
+	SvcName           string
+	timeoutsTableName string
+	exit              chan bool
+}
+
+func (tm *TimeoutManager) ensureSchema() error {
+	tblName := tm.timeoutsTableName
+	tx, e := tm.Txp.New()
+	if e != nil {
+		tm.Log().WithError(e).Error("failed to create schema for mysql timeout manager")
+		return e
+	}
+
+	createTableSQL := `CREATE TABLE IF NOT EXISTS ` + tblName + ` (
+      rec_id INT PRIMARY KEY AUTO_INCREMENT,
+      saga_id VARCHAR(255) UNIQUE NOT NULL,
+	  timeout DATETIME NOT NULL,
+	  INDEX (timeout),
+	  INDEX (saga_id)
+	 )`
+
+	if _, e := tx.Exec(createTableSQL); e != nil {
+		if rbkErr := tx.Rollback(); rbkErr != nil {
+			tm.Log().Warn("timeout manager failed to rollback transaction")
+		}
+		return e
+	}
+	return tx.Commit()
+}
+
+func (tm *TimeoutManager) purge() error {
+	purgeSQL := `DELETE FROM ` + tm.timeoutsTableName
+
+	tx, e := tm.Txp.New()
+	if e != nil {
+		tm.Log().WithError(e).Error("failed to purge timeout manager")
+		return e
+	}
+
+	if _, execErr := tx.Exec(purgeSQL); execErr != nil {
+		tm.Log().WithError(execErr).Error("failed to purge timeout manager")
+		return tx.Rollback()
+	}
+	return tx.Commit()
+}
+
+//Start starts the timeout manager
+func (tm *TimeoutManager) Start() error {
+	go tm.trackTimeouts()
+	return nil
+}
+
+//Stop shuts down the timeout manager
+func (tm *TimeoutManager) Stop() error {
+	tm.exit <- true
+	return nil
+}
+
+func (tm *TimeoutManager) trackTimeouts() {
+	tick := time.NewTicker(time.Second * 1).C
+	for {
+		select {
+		case <-tick:
+			tx, txe := tm.Txp.New()
+			if txe != nil {
+				tm.Log().WithError(txe).Warn("timeout manager failed to create a transaction")
+				continue
+			}
+			now := time.Now().UTC()
+			getTimeoutsSQL := `select saga_id from ` + tm.timeoutsTableName + ` where timeout < ? LIMIT 100`
+			rows, selectErr := tx.Query(getTimeoutsSQL, now)
+			if selectErr != nil {
+				tm.Log().WithError(selectErr).Error("timeout manager failed to query for pending timeouts")
+				rows.Close()
+				continue
+			}
+
+			sagaIDs := make([]string, 0)
+			for rows.Next() {
+
+				var sagaID string
+
+				if err := rows.Scan(&sagaID); err != nil {
+					tm.Log().WithError(err).Error("failed to scan timeout record")
+				}
+				sagaIDs = append(sagaIDs, sagaID)
+			}
+			tm.executeTimeout(sagaIDs)
+		case <-tm.exit:
+			return
+		}
+	}
+}
+
+func (tm *TimeoutManager) lockTimeoutRecord(tx *sql.Tx, sagaID string) error {
+
+	selectTimeout := `SELECT saga_id FROM ` + tm.timeoutsTableName + ` WHERE saga_id = ? FOR UPDATE`
```

Comment:
You should add here skip locked

Reply:
why ? I am selecting by a unique key and not doing a range query so I do not think the skip locked makes a difference here